src/doc_handle.rs

```diff
@@ -0,0 +1,46 @@
+use automerge as am;
+use futures::Stream;
+
+use crate::DocumentId;
+
+/// A handle to a document managed by the repository
+///
+/// This is effectively an async mutex around an automerge document
+#[derive(Clone)]
+pub struct DocHandle {}
+
+impl DocHandle {
+    pub fn id(&self) -> DocumentId {
+        todo!()
+    }
+
+    /// Wait for this document to be unlocked and then modify it
+    ///
+    /// Any changes made to the document by `F` will be enqueue to send to all
+    /// connected peers who are sharing this document
+    pub async fn modify<O, E, F: FnOnce(&am::Automerge) -> Result<O, E>>(
```

Comment:
I would rather not have the methods of the handle be async, because it seems that a doc handle doesn't have much to do with async/await, a bit like the underlying document also isn't async. 

On the other hande an async version of the current `wait_read`, or `with_doc_mut`, could be useful to avoid people having to use a `spawn_blocking` like API if they want to block on the document becoming ready in an async context.

Reply:
Note: I removed the "document ready" concept from the handle at https://github.com/automerge/spanreed/pull/11

Essentially I changed it so that for a bootstrapped document, one gets a handle via the sync observer, and only once it's ready.